### ISSUE TITLE
Alternative to `secrecy`

### ISSUE THREAD

**Issue by radumarias** [xoriors/rencfs#44]
#update we migrated to https://github.com/Eyob94/shush-rs which still needs to implement `mlock` and `mprotect` and maybe expiry.
So normally this should be closed, keeping it open until the above is implemented.

Use [secrets](https://crates.io/crates/secrets) a`secrecy` doesn't have `mlock` nor `mprotect`, that one does.

Alternatives, investigate if these makes sense to be used instead of `secrets`
- https://crates.io/crates/memsecurity
- https://docs.rs/clear_on_drop/0.2.4/clear_on_drop/

**Comment by Eyob94** [comment]
Based on my research so far

1. `secrets` crate
- Supports both mlock and mprotect(for heap-allocated buffers)
- Zero's out memory when no longer in use
- Well compatible with libsodium
- No mention of audits
- Not panic safe, vulnerable to segmentation faults in case of violation of mprotect's protection level
- Mostly a wrapper around Sodium

2. `memsecurity` crate
- Supports mlock ONLY for the encryption key that will be used to encrypt the text to be protected
- Uses Ascon 128a cipher under the hood
- No mention of mprotect
- Zero's out memory when no longer in use
- Not Audited

3. `clear_on_drop` crate
- Used only for Zeroizing out values
- Requires an additional library to handle encryption
- No mlock or mprotect used, must be coupled with another library

Another option would be to use the [`memsec`](https://docs.rs/memsec/latest/memsec/index.html) library. It has the cross-platform implementations for both `mlock`,`mprotect` and `memzero`. It's also the library used by `memsecurity` under the hood.

**Comment by radumarias** [comment]
very good analysis, I'm impressed :)
I would go with secrets as seems it offers the most of security with both
mprotect and mlock

my only concern is if we can compile it on wasm, android and iOS as it's
using libsodium as C bindings, could you try to compile on those targets to
see how it goes?

**Comment by radumarias** [comment]
I used another one that has mlock that's minimal and most likely compiles
on all targets
https://crates.io/crates/memsec

please give it a try and compile on those targets

it would mean to implement our own secrecy + memsec, but we could fork
secrecy and add this

we could then propose a new crate, as in secrecy is hard to do PR as I
tried and it's no_std and this is std as I known, but check it out

**Comment by Eyob94** [comment]
Thanks! I also leaned towards `memsec`, it is what's used under `memsecurity` after all as I mentioned it in the analysis. I'll take a look and let you know

**Comment by radumarias** [comment]
memsecurity sounds good, please check if they compile on wasm, Android and
iOS targets.
if so, I think that's the one for us

thank you for making all of these, I really appreciate it

**Comment by radumarias** [comment]
please install these targets and see which of the above compiles ok, I expect to have issues with `libsodium` on mobile

```
wasm32-unknown-unknown
wasm32-wasi
aarch64-linux-android
armv7-linux-androideabi
i686-linux-android
aarch64-apple-ios
aarch64-apple-ios-sim
x86_64-apple-ios
i686-pc-windows-gnu
x86_64-pc-windows-gnu
i686-unknown-linux-gnu
x86_64-unknown-linux-gnu
x86_64-unknown-linux-musl
aarch64-apple-darwin
x86_64-apple-darwin
```

**Comment by Eyob94** [comment]
Results

| Targets                          | Secrets      | MemSecurity  | Memsec       |
|----------------------------------|--------------|--------------|--------------|
| wasm32-unknown-unknown           | fail         | fail         | fail         |
| wasm32-wasi                      | fail         | fail         | fail         |
| aarch64-linux-android            | fail         | fail         | fail         |
| armv7-linux-androideabi          | fail         | fail         | fail         |
| i686-linux-android               | fail         | fail         | fail         |
| aarch64-apple-ios                | fail         | fail         | fail         |
| aarch64-apple-ios-sim            | fail         | fail         | fail         |
| x86_64-apple-ios                 | fail         | fail         | fail         |
| i686-pc-windows-gnu              | fail         | fail         | fail         |
| x86_64-pc-windows-gnu            | fail         | fail         | fail         |
| i686-unknown-linux-gnu           | fail         | fail         | fail         |
| x86_64-unknown-linux-gnu         | success      | success      | success      |
| x86_64-unknown-linux-musl        | fail         | success      | success      |
| aarch64-apple-darwin             | fail         | fail         | fail         |
| x86_64-apple-darwin              | fail         | fail         | fail         |

**Comment by Eyob94** [comment]
Here's the script used

```bash
#!/bin/bash

set -euo pipefail

secret_crates=("secrets" "memsecurity" "memsec")

binary_targets=(
    "wasm32-unknown-unknown" "wasm32-wasi" "aarch64-linux-android" 
    "armv7-linux-androideabi" "i686-linux-android" "aarch64-apple-ios"
    "aarch64-apple-ios-sim" "x86_64-apple-ios" "i686-pc-windows-gnu"
    "x86_64-pc-windows-gnu" "i686-unknown-linux-gnu" "x86_64-unknown-linux-gnu"
    "x86_64-unknown-linux-musl" "aarch64-apple-darwin" "x86_64-apple-darwin"
)

rm -rf "crate_build_results.md"
touch "crate_build_results.md"

printf "| %-32s | %-12s | %-12s | %-12s |\n" "Targets" "Secrets" "MemSecurity" "Memsec" >> "crate_build_results.md"
printf "|----------------------------------|--------------|--------------|--------------|\n" >> "crate_build_results.md"

for target in "${binary_targets[@]}"
do
    echo "Adding target: $target"
    if rustup target add "$target"; then
        printf "| %-32s " "$target" >> "crate_build_results.md"
        echo "Target $target added successfully."
    else
        echo "Failed to add target $target!" >&2
        continue
    fi


    for crate in "${secret_crates[@]}"
    do
        if cargo add "$crate"; then
            echo "Crate $crate added successfully."

            cargo clean
            if cargo build --target "$target"; then
                echo "Build for $target completed successfully."
                build_status="success"
            else
                echo "Build for $target failed!" >&2
                build_status="fail"
            fi

            printf "| %-12s " "$build_status" >> "crate_build_results.md"
        else
            echo "Failed to add crate $crate!" >&2
            exit 1
        fi


        sleep 2
        if cargo remove "$crate"; then
            echo "$crate removed successfully"
        else
            echo "Error removing crate. Aborting"
            exit 1
        fi
    done

    sleep 2

    printf "|\n" >> "crate_build_results.md"

done

```

**Comment by radumarias** [comment]
hmm, that looks bad, but it's possible that other crates are the cause. Create a new empty project and add each of the crates individually and see then.

**Comment by Eyob94** [comment]
> hmm, that looks bad, but it's possible that other crates are the cause. Create a new empty project and add each of the crates individually and see then.

Yes that's what I did. Feel free to create a new crate, copy the script inside that crate and just run it

**Comment by Eyob94** [comment]
Here's an updated script, it creates a new crate and adds the other packages and tests on every target. Should take 10-20 minutes MAX

```bash
#!/bin/bash

set -euo pipefail

rm -rf crate_build_test

cargo new crate_build_test
cd crate_build_test

secret_crates=("secrets" "memsecurity" "memsec")

binary_targets=(
    "wasm32-unknown-unknown" "wasm32-wasi" "aarch64-linux-android" 
    "armv7-linux-androideabi" "i686-linux-android" "aarch64-apple-ios"
    "aarch64-apple-ios-sim" "x86_64-apple-ios" "i686-pc-windows-gnu"
    "x86_64-pc-windows-gnu" "i686-unknown-linux-gnu" "x86_64-unknown-linux-gnu"
    "x86_64-unknown-linux-musl" "aarch64-apple-darwin" "x86_64-apple-darwin"
)

rm -rf "crate_build_results.md"
touch "crate_build_results.md"

printf "| %-32s | %-12s | %-12s | %-12s |\n" "Targets" "Secrets" "MemSecurity" "Memsec" >> "crate_build_results.md"
printf "|----------------------------------|--------------|--------------|--------------|\n" >> "crate_build_results.md"

for target in "${binary_targets[@]}"
do
    echo "Adding target: $target"
    if rustup target add "$target"; then
        printf "| %-32s " "$target" >> "crate_build_results.md"
        echo "Target $target added successfully."
    else
        echo "Failed to add target $target!" >&2
        continue
    fi


    for crate in "${secret_crates[@]}"
    do
        if cargo add "$crate"; then
            echo "Crate $crate added successfully."

            cargo clean
            if cargo build --target "$target"; then
                echo "Build for $target completed successfully."
                build_status="success"
            else
                echo "Build for $target failed!" >&2
                build_status="fail"
            fi

            printf "| %-12s " "$build_status" >> "crate_build_results.md"
        else
            echo "Failed to add crate $crate!" >&2
            exit 1
        fi


        sleep 2
        if cargo remove "$crate"; then
            echo "$crate removed successfully"
        else
            echo "Error removing crate. Aborting"
            exit 1
        fi
    done

    sleep 2

    printf "|\n" >> "crate_build_results.md"

done

```

**Comment by radumarias** [comment]
Right, sorry didn't read well enough. How about secrecy that we use now? I expect that will be ok for much many more.

Strange for `memsec`, because I'm using it here `https://github.com/radumarias/zeroize-python/blob/main/Cargo.toml#L27` and I do macos and windows builds which works fine https://github.com/radumarias/zeroize-python/blob/main/.github/workflows/ci.yml#L36

Maybe I sent you wrong targets for at least these 2 OSs?

We need something to work on mobile too and ideally also on wasm as plan is to use this as lib on such platforms.

**Comment by radumarias** [comment]
valid targets https://chatgpt.com/share/6cf35dec-b12c-41ad-9be7-b5d3339e17c3

**Comment by radumarias** [comment]
I recommend this
- fork https://github.com/iqlusioninc/crates/tree/main/secrecy, they simplified now it warps a `Box<Zeroize>`
- with target os guards use `memsec` to `mlock` on targets where it works
- on unsupported targets we will see in the future how we can implement those, maybe we need to call other native C functions

**Comment by radumarias** [comment]
@Eyob94 do you still need to implement something in shush-rs? We said to keep this open until you finish those.
